### DISCUSSION TITLE
The correct WWDR certificates

### DISCUSSION THREAD

**Discussion by socsieng** [alexandercerutti/passkit-generator#87] [General]
I ran into an issue with certificates earlier. Specifically the WWDR certificates.

I was under the impression that any of the WWDR certificates from https://www.apple.com/certificateauthority/ could be used (I was using G3) and I couldn't figure out why the pass wasn't being loaded in the simulator and my device.

I found the following tool https://pkpassvalidator.azurewebsites.net which indicated that there was an issue with _WWDR Certificate Subject_. After switching to the G1 certificate, it ended up working.

I'm not sure if anyone else has run into this issue, with different WWDR certificates. If so, it might be a good idea to update the [documentation](https://github.com/alexandercerutti/passkit-generator/wiki/Generating-Certificates#completely-understanding-the-goal) to reflect this.

**Comment by alexandercerutti** [comment]
Uh, that's interesting. I never understood the difference between G1 and G3 and so on, even if I tried to search for something about them. Maybe it is related to something that changes in their creation or the authority, I don't know. Do you have any ideas?

I previously had some reports about bad certificates, but I wasn't able to replicate them. Maybe that was the issue. 🤔 Nice to know. That might be worth being explicated in the certificates guide, as you say. But first I'd like to understand something more, so we can provide more information. Don't you agree?

Thank you for your precious contribution!

**Reply by socsieng** [reply]
> Uh, that's interesting. I never understood the difference between G1 and G3 and so on, even if I tried to search for something about them. Maybe it is related to something that changes in their creation or the authority, I don't know. Do you have any ideas?

I don't have any ideas why unfortunately. I assumed that G1, G3, or the latest G6 would all work. However, it only seemed to work when switching from G3 to G1. I couldn't find any logs to find the actual problem so I'm concluding "it didn't work with G3, when I switched to G1, and it did work, therefore the problem is G3"

> But first I'd like to understand something more, so we can provide more information. Don't you agree?

I wasn't very thorough in my testing so I was primarily motivated by getting it to work. The correct approach definitely involves a better understanding.

**Reply by alexandercerutti** [reply]
I tried to analyze all the available WWDR certificates with this command:

```sh
$ openssl x509 -inform der -in AppleWWDRCAG3.cer -noout -text
```

What I see as a major difference is that:

- G1 is based upon `sha1`

![image](https://user-images.githubusercontent.com/5442545/139962047-b4a65a10-037f-460d-a879-924815abe54c.png)

- G2 is based upon `ecdsa` with `sha256`

![image](https://user-images.githubusercontent.com/5442545/139962133-e149135d-3724-4d69-be78-df9524a4b723.png)

- G3 is based upon `sha256`

![image](https://user-images.githubusercontent.com/5442545/139962178-e07d0b80-c899-4662-a0bb-cd22f96072af.png)

- G5, like G3, is based upon `sha256`

![image](https://user-images.githubusercontent.com/5442545/139962374-78b0c757-1920-4e0a-b857-44b18279a241.png)

- G6, like G2, is based upon `ecdsa` with (instead) `sha384`.

![image](https://user-images.githubusercontent.com/5442545/139965761-3670830a-5912-469f-ad2b-ad5b6d3bbcc0.png)

So, by excluding G2 and G6 (maybe they could also be used, but they would require a different usage of node-forge that I don't know how to implement), we are left with G1, G3, and G5.

So, you said that G1 seems to be the only one working... but my thoughts point to this line:

https://github.com/alexandercerutti/passkit-generator/blob/master/src/Signature.ts#L54

I'm not sure, but the issue might be that a sha-1 digest algorithm is being used and that might be the reason for G3 and G5 not working.

I have to do a few more tests by changing the digest algorithm to 256 and try to see if I'm right (even if I have a few doubts).

If so, we are not supporting more recent digest algorithms and, therefore, this library might iphotetically break in the next years. If my throughs reveal themselves to be correct, it might be worth trying to attempt two possible ways:

- An additional certificates param (that defaults to sha-1, for retro-compatibility) to specify the digest algorithm
- Trying to parse the wwdr certificate first.

What do you think about this?

**Reply by alexandercerutti** [reply]
Another thought that is coming to my mind, is about the field `OU` in the `Subject` for each certificate (Organizational Unit). What if Apple Wallet checks that and sees that it doesn't match with `Apple Worldwide Developer Relations`? 🤔 There are some tests to be done here.

**Reply by socsieng** [reply]
I think the analysis you've done is great. My thoughts are that the library should strive to support _all_ of certs provided by WWDR. If there's a technical reason for why they can't be supported, the API and the documentation should be clear about what is/isn't supported.

**Reply by alexandercerutti** [reply]
I've been digging a bit further down in this mess of... knowledge!

After having analyzed the CertificateSigningRequest and checked if its digest algorithm was something interesting regarding this topic (spoiler, it wasn't), I've found a thing: the generated certificate has an `Authority Key Identifier` that matches **ONLY WITH G1**.

![image](https://user-images.githubusercontent.com/5442545/140208865-75dae9d5-a690-4cef-b438-133ec4c6cb29.png)

I guess this is the reason for the message in Console.app:

> Invalid data error reading pass <identifier>. The passTypeIdentifier or teamIdentifier provided may not match your certificate, or the certificate trust chain could not be verified

Especially the last part, _the certificate trust chain could not be verified_. Because the certificate is generated with G1 and Apple expects G1 to verify that everything is okay.

So, I think I might add another paragraph and more details about this in the wiki.

Do you have any thoughts about this?

**Reply by socsieng** [reply]
No additional thoughts. I think updating the wiki to refer to the documentation would be the best course of action.

One other thing to consider might be to include the certificate in the API. This would reduce the number of steps required in the certificate setup process, reduce the number of parameters developers would need to supply, and reduce the number of potential causes of failure.

This is what a [pass-js](https://github.com/tinovyatkin/pass-js#get-your-certificates) does:

> The Apple Worldwide Developer Relations Certification Authority certificate is not needed anymore since it is already included in this package.

**Reply by alexandercerutti** [reply]
I always refused to include it inside the package because it might change in the next years and I'd like this package to keep working even if, one day, I'll stop maintaining it actively. Honestly, I don't want to release a major just to change something that is not in passkit-generator scope. WWDR Certificate is something in Apple scope that also doesn't even depend on Apple Wallet Passes API.

**Reply by socsieng** [reply]
That's fair enough. I understand this motivation.

The flip side to this is that when the certificate does expire in 2023, it may be easier for your consumers to update the package than having to download the new certificate and converting them to `.pem`.

Come certificate expiry (with BYO certificate), it would be very important to make sure that the error messages are clear about this and how to resolve it.

**Reply by alexandercerutti** [reply]
But that would be a one-time operation for them, hypothetically for the rest of their project. I think that a WWDR certificate might last over a decade.

This is G1:
![image](https://user-images.githubusercontent.com/5442545/140232089-c88db370-3aa4-495b-b2f4-61572e2d0cab.png)

Even if the next certificate will still last a decade, a major version should be released within 1 year and a few months and then nothing else, but, honestly, I cannot predict where I'll be in a year or in ten years 😂

Also, another point of view is that I would like people to learn new things, just like I do every day, but by explaining them in the right way. I might be considered unpleasant or sadistic, but I really think that people learn things by "suffering" (emotional pain) and, moreover, by **doing things**.

About explicating in case of expiration, I think that analyzing certificates to seek for expiration, would lead to a useless overhead in the library.

As I said to you in another Discussion thread, Apple Wallet validates itself the passes and logs are printed inside Console.app on Mac, so it should be used as a companion for debugging. The only thing that I don't know is if it actually prints something different if the certificate is expired or if the message is too generic.

If it is too generic, I might think of introducing such overhead for checking.
